src/ray/raylet/node_manager.cc

```diff
@@ -1339,6 +1311,45 @@ void NodeManager::ProcessAnnounceWorkerPortMessage(
   if (is_worker) {
     worker_pool_.OnWorkerStarted(worker);
     HandleWorkerAvailable(worker);
+  } else {
+    // Driver is ready. Add the job to GCS.
+    JobID job_id = worker->GetAssignedJobId();
+    boost::optional<const rpc::JobConfig &> job_config =
+        worker_pool_.GetJobConfig(job_id);
+    RAY_CHECK(job_config.has_value());
+
+    rpc::Address driver_address;
+    // Assume raylet ID is the same as the node ID.
+    driver_address.set_raylet_id(self_node_id_.Binary());
+    driver_address.set_ip_address(worker->IpAddress());
+    driver_address.set_port(port);
+    driver_address.set_worker_id(worker->WorkerId().Binary());
+    auto job_data_ptr =
+        gcs::CreateJobTableData(job_id,
+                                /*is_dead=*/false,
+                                driver_address,
+                                worker->GetProcess().GetId(),
+                                string_from_flatbuf(*message->entrypoint()),
+                                *job_config);
+
+    RAY_CHECK_OK(gcs_client_->Jobs().AsyncAdd(job_data_ptr, [client](Status status) {
+      if (!status.ok()) {
+        RAY_LOG(ERROR) << "Failed to add job to GCS: " << status.ToString();
+      }
+      // Write the reply back.
+      flatbuffers::FlatBufferBuilder fbb;
+      auto message = protocol::CreateAnnounceWorkerPortReply(fbb);
+      fbb.Finish(message);
+
+      auto reply_status = client->WriteMessage(
```

Comment:
Why don't we call `WriteMessageAsync` here following

```
auto reply =
        ray::protocol::CreateRegisterClientReply(fbb,
                                                 status.ok(),
                                                 fbb.CreateString(status.ToString()),
                                                 to_flatbuf(fbb, self_node_id_),
                                                 assigned_port);
    fbb.Finish(reply);
    client->WriteMessageAsync(
        static_cast<int64_t>(protocol::MessageType::RegisterClientReply),
        fbb.GetSize(),
        fbb.GetBufferPointer(),
        [this, client](const ray::Status &status) {
          if (!status.ok()) {
            DisconnectClient(client,
                             rpc::WorkerExitType::SYSTEM_ERROR,
                             "Worker is failed because the raylet couldn't reply the "
                             "registration request.");
          }
        });
```

Reply:
makes sense. updated